File: flink/src/main/scala/com/clarence/dajiangtai/course01/scala/WordCount.scala
package com.clarence.dajiangtai.course01.scala

import com.clarence.dajiangtai.course01.java.dataset.WordCountData
import org.apache.flink.api.scala.ExecutionEnvironment
import org.apache.flink.api.scala._
object WordCount {

  def main(args: Array[String]): Unit = {
    // 1.获取ENV
    val env = ExecutionEnvironment.getExecutionEnvironment

    // 2. 定义Source
    val rowData = env.fromCollection(WordCountData.WORDS)
    rowData.print()


    // 3. DataSet Function
    // map: 一进一出
    // flatmap: 一进多出
    // 分词->分区->计数
    val counts = rowData.flatMap(_.toLowerCase.split("\\W+"))
        .map((_, 1))
        .groupBy(0)
        .sum(1)

    counts.print()

    // 4. Run JOB
    env.execute("WordCount")
  }



}

File: flink/src/main/scala/com/clarence/selflearn/WordCount.java
package com.clarence.selflearn;

import lombok.extern.slf4j.Slf4j;
import org.apache.flink.api.common.functions.FlatMapFunction;
import org.apache.flink.api.java.ExecutionEnvironment;
import org.apache.flink.api.java.operators.DataSource;
import org.apache.flink.api.java.tuple.Tuple2;
import org.apache.flink.util.Collector;
import scala.tools.nsc.interactive.Lexer;

@Slf4j
public class WordCount {
    public static final String WORD_COUNT_TXT_PATH = "E:\\Project\\DataSource\\wordcount.txt";

    public static void main(String[] args) throws Exception{
        // 1.ENV
        ExecutionEnvironment env = ExecutionEnvironment.getExecutionEnvironment();

        // 2.source
        DataSource<String> dataSource = env.readTextFile(WORD_COUNT_TXT_PATH);

        // 3.Function
        dataSource.flatMap(new Tokenizer()).print();


        // 3.execute
    }


    public static class Tokenizer implements FlatMapFunction<String, Tuple2<String, Long>> {

        @Override
        public void flatMap(String in, Collector<Tuple2<String, Long>> out) throws Exception {
            String[] split = in.split("\\W+");
            for (String s : split) {
                out.collect(new Tuple2<String, Long>(s, 1l));
            }
        }
    }
}
